lightningd/pay.c

```diff
@@ -1070,3 +1071,103 @@ static const struct json_command listsendpays_command = {
 	"Show sendpay, old and current, optionally limiting to {bolt11} or {payment_hash}."
 };
 AUTODATA(json_command, &listsendpays_command);
+
+static struct command_result *json_createonion(struct command *cmd,
+						const char *buffer,
+						const jsmntok_t *obj UNNEEDED,
+						const jsmntok_t *params)
+{
+	const jsmntok_t *hopstok, *hop, *payloadtok, *typetok, *pubkeytok;
+	enum sphinx_payload_type type;
+	size_t i, hop_count = 0;
+	struct json_stream *response;
+	struct pubkey pubkey;
+	struct secret *session_key, *shared_secrets;
+	struct sphinx_path *sp;
+	u8 *assocdata, *payload, *serialized;
+	struct onionpacket *packet;
+
+	if (!param(cmd, buffer, params,
+		   p_req("hops", param_array, &hopstok),
+		   p_req("assocdata", param_bin_from_hex, &assocdata),
+		   p_opt("session_key", param_secret, &session_key),
+		   NULL)) {
+		return command_param_failed();
+	}
+
+	if (session_key == NULL)
+		sp = sphinx_path_new(cmd, assocdata);
+	else
+		sp = sphinx_path_new_with_key(cmd, assocdata, session_key);
+
+	json_for_each_arr(i, hop, hopstok) {
+		payloadtok = json_get_member(buffer, hop, "payload");
+		typetok = json_get_member(buffer, hop, "type");
+		pubkeytok = json_get_member(buffer, hop, "pubkey");
+
+		if (!pubkeytok)
+			return command_fail(cmd, JSONRPC2_INVALID_REQUEST,
+					    "Hop %zu does not have a pubkey", i);
+
+		if (!payloadtok)
+			return command_fail(cmd, JSONRPC2_INVALID_REQUEST,
+					    "Hop %zu does not have a payload", i);
+
+		payload = json_tok_bin_from_hex(cmd, buffer, payloadtok);
+		if (!json_to_pubkey(buffer, pubkeytok, &pubkey))
+			return command_fail(
+			    cmd, JSONRPC2_INVALID_PARAMS,
+			    "'pubkey' should be a pubkey, not '%.*s'",
+			    pubkeytok->end - pubkeytok->start,
+			    buffer + pubkeytok->start);
+
+		if (!payload)
+			return command_fail(
+			    cmd, JSONRPC2_INVALID_PARAMS,
+			    "'payload' should be a hex encoded binary, not '%.*s'",
+			    pubkeytok->end - pubkeytok->start,
+			    buffer + pubkeytok->start);
+
+		
+		if (!typetok || !json_tok_streq(buffer, typetok, "legacy")) {
```

Comment:
Prefer an explicit test here?  Either "legacy" or "raw" (the default), rather than a random string.  Particularly if we add "tlv" later.

Reply:
There is actually no difference between raw and tlv:

https://github.com/ElementsProject/lightning/blob/21555b277d2891c9b306e677f77441c81b068549/common/sphinx.c#L526-L530

Raw really is just confusing and we should remove it.

Reply:
We already have an enum for that (`enum sphinx_payload_type`) so I went ahead and just added an explicit check for `legacy` or `tlv`, and added an error message if neither is used,